### PR TITLE
[Fix] csr_matrix_times_vector indexing 

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

This fixes `csr_matrix_times_vector` to use 1 indexed `u` and `v` for the CSR matrix format. It also updates the reverse mode version to support `var` types on the left and right hand side.

## Tests

Tests are updated to change for 0 indexed values, the sizes of `u` and `v`, and correctness with the `expect_ad()` testing framework.

```
./runTests.py test/unit/math/mix/fun/csr_matrix_times_vector_test.cpp
```

## Release notes

Fixes `csr_matrix_times_vector` indexing

## Checklist

- [x] Issue: Fixes https://github.com/stan-dev/stan/issues/3057

- [x] Copyright holder: Steve Bronder

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/prim/fun/csr_matrix_times_vector.hpp [hunk 1]
```diff
@@ -70,42 +70,32 @@ namespace math {
...
+          require_all_not_rev_matrix_t<T1, T2>* = nullptr>
 inline Eigen::Matrix<return_type_t<T1, T2>, Eigen::Dynamic, 1>
 csr_matrix_times_vector(int m, int n, const T1& w, const std::vector<int>& v,
                         const std::vector<int>& u, const T2& b) {
-  using result_t = return_type_t<T1, T2>;
-
   check_positive("csr_matrix_times_vector", "m", m);
   check_positive("csr_matrix_times_vector", "n", n);
   check_size_match("csr_matrix_times_vector", "n", n, "b", b.size());
+  check_size_match("csr_matrix_times_vector", "w", w.size(), "v", v.size());
```
Comment (alyst) [inline stan/math/prim/fun/csr_matrix_times_vector.hpp:80]: This check is already there 2 lines below.

File: stan/math/rev/fun/csr_matrix_times_vector.hpp [hunk 2]
```diff
@@ -40,39 +74,75 @@ namespace math {
...
+  using sparse_val_mat
+      = Eigen::Map<const Eigen::SparseMatrix<double, Eigen::RowMajor>>;
+  using sparse_dense_mul_type
+      = decltype((std::declval<sparse_val_mat>() * value_of(b)).eval());
+  using return_t = return_var_matrix_t<sparse_dense_mul_type, T1, T2>;
+
   check_positive("csr_matrix_times_vector", "m", m);
   check_positive("csr_matrix_times_vector", "n", n);
   check_size_match("csr_matrix_times_vector", "n", n, "b", b.size());
+  check_size_match("csr_matrix_times_vector", "w", w.size(), "v", v.size());
```
Comment (alyst) [inline stan/math/rev/fun/csr_matrix_times_vector.hpp:90]: This check is already there 2 lines below.

File: stan/math/prim/fun/csr_matrix_times_vector.hpp [hunk 3]
```diff
@@ -70,42 +70,32 @@ namespace math {
...
-         ++nze, ++i) {
-      check_range("csr_matrix_times_vector", "j", n, v[nze]);
-      b_sub.coeffRef(i) = b.coeff(v[nze] - stan::error_index::value);
-    }
-    Eigen::Matrix<value_type_t<T1>, Eigen::Dynamic, 1> w_sub(
-        w.segment(u[row] - stan::error_index::value, idx));
-    result.coeffRef(row) = dot_product(w_sub, b_sub);
-  }
-  return result;
+  std::vector<int> v_zeroed(v.size());
```
Comment (alyst) [inline stan/math/prim/fun/csr_matrix_times_vector.hpp:88]: "zeroed" sounds like "set to 0". Maybe `v_0based` or even `v0`?

Comment (SteveBronder) [reply stan/math/prim/fun/csr_matrix_times_vector.hpp:88]: How about `*_zero_based`? imo I like longer names

**Review by rok-cesnovar — APPROVED** [review]
Looks good to me.

**Pull request merged**